File: web/src/api/devices.ts
import type { LocalAccount } from "../storage/keyStore";
import type { IdentityBundle } from "../crypto/identity";
import { signedFetch } from "./signedRequest";
import { identityBundleToJson } from "./codec";
import { API_BASE } from "./register";

export interface DeviceInfo {
  id: string;
  label: string;
  createdAt: string;
}

/** No ownership check server-side: also used to discover a contact's devices for fan-out. */
export async function listDevices(accountId: string, account: LocalAccount): Promise<DeviceInfo[]> {
  const response = await signedFetch(`/v1/accounts/${accountId}/devices`, "GET", account);
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: "failed to list devices" }));
    throw new Error(error.error ?? "failed to list devices");
  }
  const raw = (await response.json()) as { id: string; label: string; created_at: string }[];
  return raw.map((d) => ({ id: d.id, label: d.label, createdAt: d.created_at }));
}

export async function linkInit(account: LocalAccount): Promise<string> {
  const response = await signedFetch(`/v1/accounts/${account.accountId}/devices/link-init`, "POST", account);
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: "failed to start device link" }));
    throw new Error(error.error ?? "failed to start device link");
  }
  const { code } = (await response.json()) as { code: string };
  return code;
}

/** Unauthenticated - the new device has no credentials yet; the code is what authorizes this. */
export async function completeLink(accountId: string, code: string, label: string, identity: IdentityBundle): Promise<string> {
  const body = { code, label, ...identityBundleToJson(identity) };
  const response = await fetch(`${API_BASE}/v1/accounts/${accountId}/devices`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: "failed to complete device link" }));
    throw new Error(error.error ?? "failed to complete device link");
  }
  const { device_id } = (await response.json()) as { device_id: string };
  return device_id;
}

export async function unlinkDevice(deviceId: string, account: LocalAccount): Promise<void> {
  const response = await signedFetch(`/v1/devices/${deviceId}`, "DELETE", account);
  if (!response.ok) {
    const error = await response.json().catch(() => ({ error: "failed to unlink device" }));
    throw new Error(error.error ?? "failed to unlink device");
  }
}

File: server/src/routes/devices.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use chrono::{DateTime, Duration, Utc};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use uuid::Uuid;

use super::register::{insert_device_bundle, validate_bundle, RegisterRequest};
use crate::auth::AuthenticatedDevice;
use crate::error::{bad_request, forbidden, not_found, server_error, ApiError};

const LINK_CODE_TTL: Duration = Duration::minutes(5);

async fn require_same_account(pool: &PgPool, caller_device_id: Uuid, account_id: Uuid) -> Result<(), ApiError> {
    let caller_account_id = sqlx::query_scalar!("SELECT account_id FROM devices WHERE id = $1", caller_device_id)
        .fetch_optional(pool)
        .await
        .map_err(|_| server_error())?
        .ok_or_else(|| server_error())?;
    if caller_account_id != account_id {
        return Err(forbidden("device does not belong to this account"));
    }
    Ok(())
}

#[derive(Serialize)]
pub struct LinkInitResponse {
    pub code: String,
}

/// Only an already-authenticated device of `account_id` can start a link -
/// otherwise anyone could mint a pairing code for someone else's account.
pub async fn link_init(
    State(pool): State<PgPool>,
    AuthenticatedDevice(caller_device_id): AuthenticatedDevice,
    Path(account_id): Path<Uuid>,
) -> Result<Json<LinkInitResponse>, ApiError> {
    require_same_account(&pool, caller_device_id, account_id).await?;

    // 8 random bytes, hex-encoded: short enough to type, long enough that
    // guessing it inside a 5-minute window isn't practical.
    let mut raw = [0u8; 8];
    rand::rng().fill_bytes(&mut raw);
    let code = hex::encode(raw);

    let expires_at = Utc::now() + LINK_CODE_TTL;
    sqlx::query!(
        "INSERT INTO pending_device_links (code, account_id, expires_at) VALUES ($1, $2, $3)",
        code,
        account_id,
        expires_at,
    )
    .execute(&pool)
    .await
    .map_err(|_| server_error())?;

    Ok(Json(LinkInitResponse { code }))
}

#[derive(Deserialize)]
pub struct CompleteLinkRequest {
    pub code: String,
    pub label: String,
    #[serde(flatten)]
    pub bundle: RegisterRequest,
}

#[derive(Serialize)]
pub struct CompleteLinkResponse {
    pub device_id: Uuid,
}

/// Unauthenticated: the new device has no credentials yet. Authorization
/// comes entirely from possessing a valid, unexpired code for this account.
pub async fn complete_link(
    State(pool): State<PgPool>,
    Path(account_id): Path<Uuid>,
    Json(req): Json<CompleteLinkRequest>,
) -> Result<(StatusCode, Json<CompleteLinkResponse>), ApiError> {
    let bundle = validate_bundle(&req.bundle)?;

    let mut tx = pool.begin().await.map_err(|_| server_error())?;

    let consumed = sqlx::query!(
        "DELETE FROM pending_device_links WHERE code = $1 AND account_id = $2 AND expires_at > now() RETURNING code",
        req.code,
        account_id,
    )
    .fetch_optional(&mut *tx)
    .await
    .map_err(|_| server_error())?;
    if consumed.is_none() {
        return Err(bad_request("link code is invalid, expired, or already used"));
    }

    let device_id = sqlx::query_scalar!("INSERT INTO devices (account_id, label) VALUES ($1, $2) RETURNING id", account_id, req.label)
        .fetch_one(&mut *tx)
        .await
        .map_err(|_| server_error())?;

    insert_device_bundle(&mut tx, device_id, &bundle).await?;

    tx.commit().await.map_err(|_| server_error())?;

    Ok((StatusCode::CREATED, Json(CompleteLinkResponse { device_id })))
}

#[derive(Serialize)]
pub struct DeviceOut {
    pub id: Uuid,
    pub label: String,
    pub created_at: DateTime<Utc>,
}

/// No ownership check: a sender needs to discover a *contact's* device list
/// to fan out to it, the same visibility the prekey-bundle endpoint already
/// has. See plan.md for the accepted metadata trade-off.
pub async fn list_devices(State(pool): State<PgPool>, AuthenticatedDevice(_caller): AuthenticatedDevice, Path(account_id): Path<Uuid>) -> Result<Json<Vec<DeviceOut>>, ApiError> {
    let rows = sqlx::query!("SELECT id, label, created_at FROM devices WHERE account_id = $1 ORDER BY created_at", account_id)
        .fetch_all(&pool)
        .await
        .map_err(|_| server_error())?;

    Ok(Json(
        rows.into_iter().map(|r| DeviceOut { id: r.id, label: r.label, created_at: r.created_at }).collect(),
    ))
}

/// Same-account-restricted, unlike listing: unlinking is a privileged action.
/// A device may unlink itself.
pub async fn unlink_device(
    State(pool): State<PgPool>,
    AuthenticatedDevice(caller_device_id): AuthenticatedDevice,
    Path(device_id): Path<Uuid>,
) -> Result<StatusCode, ApiError> {
    let target_account_id = sqlx::query_scalar!("SELECT account_id FROM devices WHERE id = $1", device_id)
        .fetch_optional(&pool)
        .await
        .map_err(|_| server_error())?
        .ok_or_else(|| not_found("device not found"))?;

    require_same_account(&pool, caller_device_id, target_account_id).await?;

    sqlx::query!("DELETE FROM devices WHERE id = $1", device_id)
        .execute(&pool)
        .await
        .map_err(|_| server_error())?;

    Ok(StatusCode::NO_CONTENT)
}
